modules/stitching/src/matchers.cpp

```diff
@@ -583,7 +583,10 @@ void SurfFeaturesFinderGpu::find(InputArray image, ImageFeatures &features)
     image_.upload(image);
 
     ensureSizeIsEnough(image.size(), CV_8UC1, gray_image_);
-    cvtColor(image_, gray_image_, COLOR_BGR2GRAY);
+
+    Mat gray_image;
+    cvtColor(image, gray_image, COLOR_BGR2GRAY);
+    gray_image_.upload(gray_image);
```

Comment:
HI,

I would like to separate the issues in to two:
1. Fix build issue with opencv non-free or community modules (In this case XFeatures2D) [Build TIme]
2. Camera Parameter adjustment failing. [Run Time]

For the first one, trying to change color of a GPU mat with cvtColor definitely stops the build. So, the fix is proper.

For the second case,  CPU and GPU implementation of most of the functions can never be exactly the same due to CUDA nature. From my experience, I have found different outputs for the same set of images with CPU and GPU. This can be due to either different parameters or different implementation. Sometimes algorithms are written differently to avoid performance penalty. I believe the descriptor implementation is different between CPU and GPU. You can try with adjusting some parameters and check.

You can refer to: http://answers.opencv.org/question/72530/difference-between-surf_gpu-and-non-gpu-surf/ or http://opencv-users.1802565.n2.nabble.com/GPU-Differences-in-the-descriptors-of-GPU-and-CPU-implementations-of-SURF-td7519924.html